### PR TITLE
feat(cel): add index-mutation functions to the lists library

### LINKED ISSUE #1147: feat(cel): add index-mutation functions to the lists library (setIndex, insertAt, removeAt)
### Feature Description

**Problem Statement**:

CEL's built-in list operations are all read-only. Functions like `filter()`, `map()`, `slice()`, `reverse()`, and `distinct()` return new lists derived from an existing one, but there is no way to return a new list with a single element replaced, inserted, or removed at a specific index.

This gap is particularly noticeable when writing RGD expressions that manage ordered arrays in instance specs — for example, a `[]int` field tracking per-replica state, or a `[]string` field holding a slot allocation table. The only workaround today is to reconstruct the entire list with a `map()` comprehension and a conditional, which is verbose and hard to read:

```cel
# Replacing element at index 2 — current workaround
schema.spec.slots.map(i, i == 2 ? newValue : schema.spec.slots[i])
```

The intent is buried. The index variable `i` is reused as both a loop counter and a value, which only works because the list is `[]int`. For `[]string` or `[]dyn` it breaks entirely — `i` is the index, not the element, so you'd need:

```cel
lists.range(size(schema.spec.tags)).map(i, i == 2 ? "new-tag" : schema.spec.tags[i])
```

This is workable but noisy, and it requires knowing about `lists.range` just to perform what is conceptually a single-element update.

**Proposed Solution**:

Three new functions in `pkg/cel/library/lists.go`, alongside the existing `maps.merge` pattern:

```
lists.setIndex(list(dyn), int, dyn)   -> list(dyn)   // replace element at index
lists.insertAt(list(dyn), int, dyn)   -> list(dyn)   // insert element before index
lists.removeAt(list(dyn), int)        -> list(dyn)   // remove element at index
```

The same expressions from above become:

```cel
lists.setIndex(schema.spec.slots, 2, newValue)
lists.insertAt(schema.spec.tags, 0, "prepended")
lists.removeAt(schema.spec.items, idx)
```

All three functions are pure — they return a new list and do not mutate the input. Bounds are checked and an error is returned for out-of-range indices. `insertAt` with `index == size(list)` appends (equivalent to `list + [value]`).

**Why this belongs in kro rather than cel-go**:

`cel-go/ext` lists covers general-purpose functional operations (sort, distinct, flatten, slice). Index-mutation is a different category — it's about producing modified copies of a list at a known position, which is an application-layer concern driven by how Kubernetes resource specs are structured (ordered arrays of scalars or objects). The existing `maps.merge` in `pkg/cel/library/maps.go` is the right precedent: kro maintains its own library for operations that are specific to the resource-management domain.

**Alternatives Considered**:

- `map()` + conditional: works for simple cases but requires `lists.range` to get a counter, is verbose, and the intent is not clear to someone reading the RGD for the first time.
- `cel.bind()` + `map()`: reduces repetition but doesn't change the fundamental verbosity.
- No workaround exists for `insertAt` or `removeAt` without reconstructing the entire list manually.

**Additional Context**:

`slice()` is already available via `ext.Lists()` and covers the read-only slicing case, so it is deliberately excluded from this proposal.

I have a working implementation and am happy to send a PR if the approach looks good.

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### PR DESCRIPTION
**Pull request by pnz1990**
## What

Three new functions in `pkg/cel/library/lists.go`, alongside the existing `maps.merge` pattern. All are pure — they return a new list and do not modify the input.

**1. `pkg/cel/library/lists.go`** — new functions:

```
lists.setIndex(list(dyn), int, dyn) -> list(dyn)   // replace element at index
lists.insertAt(list(dyn), int, dyn) -> list(dyn)   // insert before index; index == size appends
lists.removeAt(list(dyn), int)      -> list(dyn)   // remove element at index
```

Bounds are checked on all three. Out-of-range index returns an error. Implemented with `types.NewRefValList` so mixed-type lists round-trip correctly.

**2. `pkg/cel/environment.go`** — register `library.Lists()` in `BaseDeclarations()`

**No overlap with existing libraries** — `cel-go/ext.Lists()` and `k8s.io/apiserver` lists library both cover read-only operations only (`slice`, `sort`, `distinct`, `flatten`, `isSorted`, `sum`, `min`, `max`, `indexOf`). None of them have mutation functions.

## Why

CEL has no way to return a new list with a single element changed at a known index. The only workaround is to reconstruct the entire list through a `map()` comprehension with a conditional — which requires knowing `lists.range` just to get a counter, and breaks for non-int element types:

```cel
# before: intent buried, breaks for list(string)
lists.range(size(schema.spec.tags)).map(i, i == 2 ? "new-tag" : schema.spec.tags[i])

# after
lists.setIndex(schema.spec.tags, 2, "new-tag")
```

There is no workaround at all for `insertAt` or `removeAt` without rebuilding the list manually element by element.

## Tests

- `pkg/cel/library/lists_test.go` — `TestListsSetIndex`, `TestListsInsertAt`, `TestListsRemoveAt`: int and string elements, bounds errors, edge cases (empty list, single-element list, append via `insertAt`)
- `TestListsComposition`: chains of operations to verify purity (input not mutated, round-trips)

Closes #1147

### CODE REVIEW

File: pkg/cel/library/lists.go [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+		// lists.setIndex(arr list(dyn), index int, value dyn) -> list(dyn)
+		cel.Function("lists.setIndex",
+			cel.Overload("lists.setIndex_list_int_dyn",
+				[]*cel.Type{listDyn, cel.IntType, cel.DynType},
+				listDyn,
+				cel.FunctionBinding(listsSetIndex),
+			),
+		),
+
+		// lists.insertAt(arr list(dyn), index int, value dyn) -> list(dyn)
```
Comment (jakobmoellerdev) [inline pkg/cel/library/lists.go]: This likely needs adjust for Typed Parameters instead of using Dyn hardcoding. Can we code them not against dyn but `cel.TypeParamType(name)`?

Comment (pnz1990) [reply pkg/cel/library/lists.go]: updated to use cel.TypeParamType("T"). also added a TestListsTypeInference test .

File: pkg/cel/library/lists.go [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+}
+
+func (l *listsLibrary) ProgramOptions() []cel.ProgramOption {
+	return nil
+}
+
+// listsSetLegacy is the original lists.set implementation, typed list(int) only.
+// Kept for backwards compatibility with existing RGDs using lists.set.
+func listsSetLegacy(args ...ref.Val) ref.Val {
+	if len(args) != 3 {
```
Comment (a-hilaly) [inline pkg/cel/library/lists.go]: This is legacy now, shall we remove it?

File: pkg/cel/library/lists.go [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+	return []cel.EnvOption{
+		// lists.set is kept for backwards compatibility with existing RGDs.
+		// It is typed list(int) only. New code should use lists.setIndex.
+		cel.Function("lists.set",
+			cel.Overload("lists.set_list_int_int",
+				[]*cel.Type{cel.ListType(cel.IntType), cel.IntType, cel.IntType},
+				cel.ListType(cel.IntType),
+				cel.FunctionBinding(listsSetLegacy),
+			),
+		),
```
Comment (a-hilaly) [inline pkg/cel/library/lists.go]: ditto, we don't need to registry legacy, as it's not legacy for kro

File: pkg/cel/library/lists.go [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+// Examples:
+//
+//	lists.removeAt([1, 2, 3], 1)   // [1, 3]
+//	lists.removeAt([1, 2, 3], 0)   // [2, 3]
+//	lists.removeAt([1, 2, 3], 2)   // [1, 2]
+func Lists() cel.EnvOption {
+	return cel.Lib(&listsLibrary{})
+}
+
+type listsLibrary struct{}
```
Comment (a-hilaly) [inline pkg/cel/library/lists.go]: Can you add the version option pattern (`ListsVersion`, `ListsOption`) for consistency with `maps.go` and `json.go`?

**Comment by pnz1990** [comment]
**Downstream rebase note (for maintainers and author)**

Once this PR merges, the following rebase chain is required:

1. **Rebase `pnz1990/kro` branch `cel-writeback-d`** onto the new upstream `main` — this branch carries additional krombat-private patches (specPatch, stateWrite, CSV library, seededInt) on top of the lists functions introduced here
2. **Update `github.com/pnz1990/krombat` backend `go.mod`** — the `replace` directive for `github.com/kubernetes-sigs/kro` is pinned to a `cel-writeback-d` commit SHA; after the rebase it must be updated to the new SHA so the backend CEL playground (issue #367) picks up the merged lists functions

Tracking issue on the consuming project: https://github.com/pnz1990/krombat/issues/367

File: pkg/cel/library/lists.go [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+
+// Lists returns a CEL library that provides index-mutation functions for lists.
+// All functions are pure — they return a new list and do not modify the input.
+//
+// # SetIndex
+//
+// Returns a new list with the element at index replaced by value.
+// Index must be in [0, size(list)).
+//
+//	lists.setIndex(list(T), int, T) -> list(T)
```
Comment (jakobmoellerdev) [inline pkg/cel/library/lists.go]: shouldnt this be `setAtIndex()` since youre not setting the actual index? Below functions are also not named ideally I think.

File: pkg/cel/library/lists.go
Comment (jakobmoellerdev) [inline pkg/cel/library/lists.go:1]: The new functions should be documented at https://kro.run/next/docs/concepts/rgd/cel-expressions#available-cel-libraries I think, I am noticing now that were probably lacking docs for existing owned functions by kro as well. @a-hilaly

File: pkg/cel/library/lists.go (hunk 5 above)
Comment (pnz1990) [reply pkg/cel/library/lists.go]: ty. renamed all functions for consistency

File: pkg/cel/library/lists.go
Comment (pnz1990) [reply pkg/cel/library/lists.go:1]: added the kro-owned index mutation library to the Available CEL Libraries
